### ISSUE TITLE
Version 2.4.1.1 is very slow starting and responding

### ISSUE THREAD

**Issue by FormelLMS** [terrymacdonald/DisplayMagician#129]
Upgraded the Version to  2.4.1.1 and deleted all Display Profiles. 
All after that is responding very slow. Getting into the Display Profiles, creating one, starting and stopping for a game
[DisplayMagician-Support-20220701-1633.zip](https://github.com/terrymacdonald/DisplayMagician/files/9029640/DisplayMagician-Support-20220701-1633.zip)
.

**Comment by terrymacdonald** [comment]
@FormelLMS 

Can you please explain what you mean by 'it is responding very slow'? Is there a large difference in the time it took DisplayMagician 2.3.1.60 to change between profiles compared to DisplayMagician 2.4.1.1? If so, how much difference in time is there? Does DisplayMagician 2.4.1.1 take 20 seconds more to change screens than 2.3.1.60 did?

Some other questions I have:
- Does the delay occur when clicking on any part of DisplayMagician (e.g. when opening the main window, or when clicking on buttons compared o when they respond?
- Does this delay only occur when you change between profiles or does it occur when running a game shortcut too? 
- Is there a difference in response time if you run a display profile shortcut from the desktop, or if you click the Apply button on the Display Profile Window?
- Does the delay happen every time, and take the same amount of time? Is it consistently that time?
- Are you on Windows 11 at the moment?
- What NVIDIA driver are you running currently? 

Checking the logs, it looks like you tried to run an Automobilista 2 game from a shortcut, but the game isn't installed at the moment:

```
2022-07-01 16:11:01.4591|TRACE|DisplayMagician.Program|Program/RunShortcutTask: Got exclusive control of the RunShortcutTask|
2022-07-01 16:11:01.4920|DEBUG|DisplayMagician.ShortcutRepository|ShortcutRepository/RunShortcut: Running the shortcut Automobilista 2.|
2022-07-01 16:11:01.4920|DEBUG|DisplayMagician.Program|ProfileRepository/ContainsProfile2: Checking if our profile repository contains a profile with UUID or Name 33ac7d5f-ba99-4b42-a72f-902551b2e054|
2022-07-01 16:11:01.4920|DEBUG|DisplayMagician.Program|ProfileRepository/ContainsProfile2: Our profile repository doesn't contain a profile with a UUID or Name 33ac7d5f-ba99-4b42-a72f-902551b2e054|
2022-07-01 16:11:01.4920|WARN|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The profile UUID 33ac7d5f-ba99-4b42-a72f-902551b2e054 isn't in the ProfileRepository|
2022-07-01 16:11:01.4920|TRACE|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: This shortcut is a Game|
2022-07-01 16:11:01.4920|TRACE|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The game library is Steam|
2022-07-01 16:11:01.4920|WARN|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The game library does not have Game ID 1066890 installed!|
2022-07-01 16:11:01.4920|ERROR|DisplayMagician.ShortcutRepository|ShortcutRepository/RunShortcut: Cannot run the shortcut Automobilista 2 as it isn't valid|

```

There is also some errors when DisplayMagician attempts to find out the TaskBarLayout for your screens:
```
2022-07-01 16:11:15.5958|TRACE|DisplayMagician.Program|WinLibrary/GetWindowsDisplayConfig: Attempting to get the Windows Taskbar Layouts.|
2022-07-01 16:11:15.5958|TRACE|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Getting the monitor coordinates from the main monitor|
2022-07-01 16:11:15.6107|TRACE|DisplayMagician.Program|TaskBarLayout/GetRegKeyValueFromDevicePath: Found regKeyValue AUS2707#5&8a3f6fe&0&UID5120 in the devicePath \\?\DISPLAY#AUS2707#5&8a3f6fe&0&UID5120#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7}.|
2022-07-01 16:11:15.6107|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found MMStuckRect3 registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\MMStuckRects3|
2022-07-01 16:11:15.6107|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen. Screen details may not be available yet in registry.|
2022-07-01 16:11:15.6107|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found StuckRect3 single display registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\StuckRects3|
2022-07-01 16:11:15.6107|ERROR|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen.|
2022-07-01 16:11:15.6107|ERROR|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Taskbar read #1 from registry didn't work.|
2022-07-01 16:11:15.6107|TRACE|DisplayMagician.Program|WinLibrary/GetWindowsDisplayConfig: We were unable to get all the Windows Taskbar Layouts! So we need to try again. Attempt 1 of 4.|
2022-07-01 16:11:20.6173|TRACE|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Getting the monitor coordinates from the main monitor|
2022-07-01 16:11:20.6173|TRACE|DisplayMagician.Program|TaskBarLayout/GetRegKeyValueFromDevicePath: Found regKeyValue AUS2707#5&8a3f6fe&0&UID5120 in the devicePath \\?\DISPLAY#AUS2707#5&8a3f6fe&0&UID5120#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7}.|
2022-07-01 16:11:20.6173|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found MMStuckRect3 registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\MMStuckRects3|
2022-07-01 16:11:20.6173|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen. Screen details may not be available yet in registry.|
2022-07-01 16:11:20.6173|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found StuckRect3 single display registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\StuckRects3|
2022-07-01 16:11:20.6173|ERROR|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen.|
2022-07-01 16:11:20.6173|ERROR|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Taskbar read #1 from registry didn't work.|
2022-07-01 16:11:20.6173|TRACE|DisplayMagician.Program|WinLibrary/GetWindowsDisplayConfig: We were unable to get all the Windows Taskbar Layouts! So we need to try again. Attempt 2 of 4.|
2022-07-01 16:11:25.6200|TRACE|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Getting the monitor coordinates from the main monitor|
2022-07-01 16:11:25.6200|TRACE|DisplayMagician.Program|TaskBarLayout/GetRegKeyValueFromDevicePath: Found regKeyValue AUS2707#5&8a3f6fe&0&UID5120 in the devicePath \\?\DISPLAY#AUS2707#5&8a3f6fe&0&UID5120#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7}.|
2022-07-01 16:11:25.6200|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found MMStuckRect3 registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\MMStuckRects3|
2022-07-01 16:11:25.6200|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen. Screen details may not be available yet in registry.|
2022-07-01 16:11:25.6200|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found StuckRect3 single display registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\StuckRects3|
2022-07-01 16:11:25.6200|ERROR|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen.|
2022-07-01 16:11:25.6200|ERROR|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Taskbar read #1 from registry didn't work.|
2022-07-01 16:11:25.6200|TRACE|DisplayMagician.Program|WinLibrary/GetWindowsDisplayConfig: We were unable to get all the Windows Taskbar Layouts! So we need to try again. Attempt 3 of 4.|
2022-07-01 16:11:28.8229|TRACE|DisplayMagician.SingleInstance|SingleInstance/executeAnActionCallback: Received data from another DisplayMagician instance: D:\Systemprogramme\DisplayMagician\DisplayMagician.exe RunShortcut 19b66ed4-86f1-4fbc-b16f-0509dbf348b7|
2022-07-01 16:11:28.8229|TRACE|DisplayMagician.SingleInstance|SingleInstance/executeAnActionCallback: Other DisplayMagician instance provided the RunShortcut command: 'RunShortcut 19b66ed4-86f1-4fbc-b16f-0509dbf348b7'|
2022-07-01 16:11:28.8229|DEBUG|DisplayMagician.Program|Program/RunShortcut: Running shortcut 19b66ed4-86f1-4fbc-b16f-0509dbf348b7|
2022-07-01 16:11:28.8229|DEBUG|DisplayMagician.Program|ProfileRepository/ContainsProfile2: Checking if our profile repository contains a profile with UUID or Name 33ac7d5f-ba99-4b42-a72f-902551b2e054|
2022-07-01 16:11:28.8229|DEBUG|DisplayMagician.Program|ProfileRepository/ContainsProfile2: Our profile repository doesn't contain a profile with a UUID or Name 33ac7d5f-ba99-4b42-a72f-902551b2e054|
2022-07-01 16:11:28.8229|WARN|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The profile UUID 33ac7d5f-ba99-4b42-a72f-902551b2e054 isn't in the ProfileRepository|
2022-07-01 16:11:28.8229|TRACE|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: This shortcut is a Game|
2022-07-01 16:11:28.8229|TRACE|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The game library is Steam|
2022-07-01 16:11:28.8229|WARN|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The game library does not have Game ID 270880 installed!|
2022-07-01 16:11:28.8229|TRACE|DisplayMagician.Program|Program/RunShortcutTask: Got exclusive control of the RunShortcutTask|
2022-07-01 16:11:28.8229|DEBUG|DisplayMagician.ShortcutRepository|ShortcutRepository/RunShortcut: Running the shortcut American Truck Simulator.|
2022-07-01 16:11:28.8229|DEBUG|DisplayMagician.Program|ProfileRepository/ContainsProfile2: Checking if our profile repository contains a profile with UUID or Name 33ac7d5f-ba99-4b42-a72f-902551b2e054|
2022-07-01 16:11:28.8229|DEBUG|DisplayMagician.Program|ProfileRepository/ContainsProfile2: Our profile repository doesn't contain a profile with a UUID or Name 33ac7d5f-ba99-4b42-a72f-902551b2e054|
2022-07-01 16:11:28.8229|WARN|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The profile UUID 33ac7d5f-ba99-4b42-a72f-902551b2e054 isn't in the ProfileRepository|
2022-07-01 16:11:28.8229|TRACE|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: This shortcut is a Game|
2022-07-01 16:11:28.8229|TRACE|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The game library is Steam|
2022-07-01 16:11:28.8229|WARN|DisplayMagician.ShortcutItem|ShortcutItem/RefreshValidity: The game library does not have Game ID 270880 installed!|
2022-07-01 16:11:28.8229|ERROR|DisplayMagician.ShortcutRepository|ShortcutRepository/RunShortcut: Cannot run the shortcut American Truck Simulator as it isn't valid|
2022-07-01 16:11:30.6266|TRACE|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Getting the monitor coordinates from the main monitor|
2022-07-01 16:11:30.6266|TRACE|DisplayMagician.Program|TaskBarLayout/GetRegKeyValueFromDevicePath: Found regKeyValue AUS2707#5&8a3f6fe&0&UID5120 in the devicePath \\?\DISPLAY#AUS2707#5&8a3f6fe&0&UID5120#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7}.|
2022-07-01 16:11:30.6266|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found MMStuckRect3 registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\MMStuckRects3|
2022-07-01 16:11:30.6266|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen. Screen details may not be available yet in registry.|
2022-07-01 16:11:30.6266|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found StuckRect3 single display registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\StuckRects3|
2022-07-01 16:11:30.6266|ERROR|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen.|
2022-07-01 16:11:30.6266|ERROR|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Taskbar read #1 from registry didn't work.|
2022-07-01 16:11:30.6266|TRACE|DisplayMagician.Program|WinLibrary/GetWindowsDisplayConfig: We were unable to get all the Windows Taskbar Layouts! So we need to try again. Attempt 4 of 4.|
2022-07-01 16:11:35.6313|TRACE|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Getting the monitor coordinates from the main monitor|
2022-07-01 16:11:35.6313|TRACE|DisplayMagician.Program|TaskBarLayout/GetRegKeyValueFromDevicePath: Found regKeyValue AUS2707#5&8a3f6fe&0&UID5120 in the devicePath \\?\DISPLAY#AUS2707#5&8a3f6fe&0&UID5120#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7}.|
2022-07-01 16:11:35.6313|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found MMStuckRect3 registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\MMStuckRects3|
2022-07-01 16:11:35.6313|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen. Screen details may not be available yet in registry.|
2022-07-01 16:11:35.6313|TRACE|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Found StuckRect3 single display registry key! Software\Microsoft\Windows\CurrentVersion\Explorer\StuckRects3|
2022-07-01 16:11:35.6313|ERROR|DisplayMagician.Program|TaskBarLayout/ReadFromRegistry: Unable to get the TaskBarStuckRectangle binary settings from AUS2707#5&8a3f6fe&0&UID5120 screen.|
2022-07-01 16:11:35.6313|ERROR|DisplayMagician.Program|TaskBarLayout/GetAllCurrentTaskBarPositions: Taskbar read #1 from registry didn't work.|
```

You can see that it tried 4 times to get your taskbar location from registry, and failed. This may be the cause of your delay. WinLibrary currently waits 5 seconds if it can't read the taskbar registry, and then it tries again. This is because based on my testing, if a screen layout changes, windows takes up to 20 seconds to update registry to record this fact. We have to wait until windows has finished 4 times before we are sure to have passed the 20 second window. 

This is likely the delay you have mentioned. I *think* that I can slightly speed this up. We only MUST to do this delay when we are recording the config (i.e. creating a new display profile), and other times it's kind of a nice to have. So I've attempted to speed this up using a 'fastScan' option for the WinLibrary GetActiveConfig function. This will enable it to only query once for the general scans of the active config, and if there is a problem getting the data it will just accept that fact and will still return quickly. But it will still take up to 20 seconds when creating a new display profile as it is REALLY important we get that data correctly.

I've built a new test version based on that assumption, and included it here:
[DisplayMagicianSetup-v2.4.1.4.zip](https://github.com/terrymacdonald/DisplayMagician/files/9032315/DisplayMagicianSetup-v2.4.1.4.zip)

Please download DisplayMagician v2.4.1.4 and install it. Please check if the slowness has disappeared, and report back.

Thanks
Terry

**Comment by FormelLMS** [comment]
Hey Terry. You where absolutely right. The trouble was while opening the display configuration. And it lasts indeed 20 seconds. I’m feeling dumb now because I hadn’t read it. 
New version works good here. Thanks for that. 

Everything I’ve tried, was working there.

**Comment by terrymacdonald** [comment]
ok, closing this issue then! The hotfix I gave you will be released as part of v2.5.0 coming out in a few weeks.
